test/integration/apiserver/apply/apply_test.go

```diff
@@ -2811,3 +2811,187 @@ spec:
+	}
+
+	assertCompleteOwnership(t, (*deployment).GetManagedFields(), "scale_test")
+
+	// Re-apply the original object, it should fail with conflict because replicas have changed
+	_, err = client.CoreV1().RESTClient().Patch(types.ApplyPatchType).
+		AbsPath("/apis/apps/v1").
+		Namespace("default").
+		Resource("deployments").
+		Name("deployment").
+		Param("fieldManager", "apply_test").
+		Body(deploymentWithReplicas(1)).
+		Do(context.TODO()).
+		Get()
+	if !apierrors.IsConflict(err) {
+		t.Fatalf("Expected conflict error but got: %v", err)
+	}
+
+	// Re-apply forcing the changes should succeed
+	_, err = client.CoreV1().RESTClient().Patch(types.ApplyPatchType).
+		AbsPath("/apis/apps/v1").
+		Namespace("default").
+		Resource("deployments").
+		Name("deployment").
+		Param("fieldManager", "apply_test").
+		Param("force", "true").
+		Body(deploymentWithReplicas(1)).
+		Do(context.TODO()).
+		Get()
+	if err != nil {
+		t.Fatalf("Error updating deployment: %v ", err)
+	}
+
+	deployment, err = client.AppsV1().Deployments("default").Get(context.TODO(), "deployment", metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Failed to retrieve object: %v", err)
+	}
+
+	if *deployment.Spec.Replicas != 1 {
+		t.Fatalf("Expected replicas to be 1, but got: %d", *deployment.Spec.Replicas)
+	}
+
+	assertCompleteOwnership(t, (*deployment).GetManagedFields(), "apply_test")
+
+	// Replace scale object
+	_, err = client.CoreV1().RESTClient().
+		Put().
+		AbsPath("/apis/apps/v1").
+		Namespace("default").
+		Resource("deployments").
+		SubResource("scale").
+		Name("deployment").
+		Param("fieldManager", "replace_test").
+		Body([]byte(`{"kind":"Scale","apiVersion":"autoscaling/v1","metadata":{"name":"deployment","namespace":"default"},"spec":{"replicas":7}}`)).
+		Do(context.TODO()).
+		Get()
+	if err != nil {
+		t.Fatalf("Error updating deployment: %v ", err)
+	}
+
+	deployment, err = client.AppsV1().Deployments("default").Get(context.TODO(), "deployment", metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Failed to retrieve object: %v", err)
+	}
+
+	if *deployment.Spec.Replicas != 7 {
+		t.Fatalf("Expected to replicas to be 7, but got: %d", *deployment.Spec.Replicas)
+	}
+
+	assertCompleteOwnership(t, (*deployment).GetManagedFields(), "replace_test")
+}
+
+func deploymentWithReplicas(replicas int) []byte {
+	return []byte(fmt.Sprintf(`{
+		"apiVersion": "apps/v1",
+		"kind": "Deployment",
+		"metadata": {
+			"name": "deployment"
+		},
+		"spec": {
+			"replicas": %d,
+			"selector": {
+				"matchLabels": {
+					 "app": "nginx"
+				}
+			},
+			"template": {
+				"metadata": {
+					"labels": {
+						"app": "nginx"
+					}
+				},
+				"spec": {
+					"containers": [{
+						"name":  "nginx",
+						"image": "nginx:latest"
+					}]
+				}
+			}
+		}
+	}`, replicas))
+}
+
+func assertCompleteOwnership(t *testing.T, managedFields []metav1.ManagedFieldsEntry, fieldManager string) {
```

Comment:
Hm I wonder if, instead of the helper, it's better to get the managed fields at an index `deployment.GetManagedFields()[i]`, assert the manager name, and assert that it contains the `.spec.replicas` field.

But for the helper:
- this probably could be named assertReplicasOwnership instead
- I think we want `if !managerSeen { t.Fatalf(...) }` at the end if `.spec.replicas` is not in managed fields at all for some reason

Reply:
IIRC I think the reason we have this helper is to ensure that a new manager is owning the replicas field, but also no other manager is, right? I am not sure using `deployment.GetManagedFields()[i]` we could check both conditions.

As for the suggestions, totally! thanks for spotting, I will make the changes